cudamapper/src/utils.cpp

```diff
@@ -115,6 +121,193 @@ void print_paf(const std::vector<Overlap>& overlaps,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15};
+
+    const size_t len = seq.length();
+
+    // encode bases as nibbles and write them into the alignment struct 2 bases at a time
+    // 4 high bits are first base, 4 low bits are second base
+    uint8_t* bam_seq_start = bam_get_seq(alignment);
+    int i                  = 0;
+    for (i; (i + 1) < len; i += 2)
+    {
+        *bam_seq_start++ = (L[static_cast<unsigned char>(seq[i])] << 4) + L[static_cast<unsigned char>(seq[i + 1])];
+    }
+    // odd number of bases, so we have one left over
+    if (i < len)
+    {
+        *bam_seq_start++ = L[static_cast<unsigned char>(seq[i])] << 4;
+    }
+
+    alignment->core.l_qseq = len;
+
+    return;
+}
+
+static inline void encode_qual(bam1_t* alignment)
+{
+    // if htslib encounters 0xff at the start of the QUAL section, it writes default value (*)
+    auto bam_qual_start = bam_get_qual(alignment);
+    *bam_qual_start     = 0xff;
+}
+
+void print_sam(const std::vector<Overlap>& overlaps,
+               const std::vector<std::string>& cigars,
+               const io::FastaParser& query_parser,
+               const io::FastaParser& target_parser,
+               OutputFormat format,
+               std::mutex& write_output_mutex,
+               int argc,
+               char* argv[])
+{
+    GW_NVTX_RANGE(profiler, "print_sam::formatting_output");
+
+    // Assume output is stdout, and assume SAM output by default
+    // Rest of the add()/write() functions choose based on samFile format
+    // use hts_close here directly, because sam_close is a macro for hts_close, which the compiler did not like
+    using samFilePtr_t = std::unique_ptr<samFile, decltype(&hts_close)>;
+
+    auto file   = (format == OutputFormat::SAM) ? samFilePtr_t(sam_open("-", "wh"), &hts_close) : samFilePtr_t(sam_open("-", "bwh"), &hts_close);
+    auto header = std::unique_ptr<bam_hdr_t, decltype(&sam_hdr_destroy)>(sam_hdr_init(), &sam_hdr_destroy);
+    int result  = 0;
+
+    if (file == nullptr)
+    {
+        fprintf(stderr, "print_sam: could not open file for writing");
+        return;
```

Comment:
Should we abort at this point or throw an exception?

Reply:
I decided to throw an exception here. It _could_ be caught by the calling application and handled. One use-case I could think of it if the disk that's being written to is full or something. Also, in the future, it's plausible that we operate directly on files instead of stdout > file, so handling which file to write to could be an argument that could be changed when an exception is encountered